aten/src/ATen/core/op_registration/op_registration_test.cpp

```diff
@@ -219,4 +220,265 @@ TEST(OperatorRegistrationTest, givenOpWithMultipleKernels_whenKernelsHaveSameDis
+    auto registry = c10::RegisterOperators().op("_test::my_op" + schema, kernel<ArgTypeTestKernel>(input, std::move(inputExpectation), std::move(output)));
+    auto op = Dispatcher::singleton().findSchema("_test::my_op", "");
+    ASSERT_TRUE(op.has_value()); // assert schema is registered
+    auto actualOutput = callOp(*op, std::move(input));
+    EXPECT_EQ(1, actualOutput.size());
+    outputExpectation(actualOutput[0]);
+  }
+
+private:
+  InputType input_;
+  std::function<void(const InputType&)> inputExpectation_;
+  OutputType output_;
+  std::string schema_;
+};
+
+TEST(OperatorRegistrationTest, testAvailableArgTypes) {
+  // primitive types
+  ArgTypeTestKernel<double>::test(
+    1.5, [] (const double& v) {EXPECT_EQ(1.5, v);},
+    2.5, [] (const IValue& v) {EXPECT_EQ(2.5, v.toDouble());});
+  ArgTypeTestKernel<double>::test(
+    1.5, [] (const double& v) {EXPECT_EQ(1.5, v);},
+    2.5, [] (const IValue& v) {EXPECT_EQ(2.5, v.toDouble());},
+    "(float a) -> float");
+  ArgTypeTestKernel<int64_t>::test(
+    1, [] (const int64_t& v) {EXPECT_EQ(1, v);},
+    2, [] (const IValue& v) {EXPECT_EQ(2, v.toInt());});
+  ArgTypeTestKernel<int64_t>::test(
+    1, [] (const int64_t& v) {EXPECT_EQ(1, v);},
+    2, [] (const IValue& v) {EXPECT_EQ(2, v.toInt());},
+    "(int a) -> int");
+  ArgTypeTestKernel<bool>::test(
+    true, [] (const bool& v) {EXPECT_EQ(true, v);},
+    false, [] (const IValue& v) {EXPECT_EQ(false, v.toBool());});
+  ArgTypeTestKernel<bool>::test(
+    true, [] (const bool& v) {EXPECT_EQ(true, v);},
+    false, [] (const IValue& v) {EXPECT_EQ(false, v.toBool());},
+    "(bool a) -> bool");
+  ArgTypeTestKernel<bool>::test(
+    false, [] (const bool& v) {EXPECT_EQ(false, v);},
+    true, [] (const IValue& v) {EXPECT_EQ(true, v.toBool());});
+  ArgTypeTestKernel<bool>::test(
+    false, [] (const bool& v) {EXPECT_EQ(false, v);},
+    true, [] (const IValue& v) {EXPECT_EQ(true, v.toBool());},
+    "(bool a) -> bool");
+  ArgTypeTestKernel<std::string>::test(
+    "string1", [] (const std::string& v) {EXPECT_EQ("string1", v);},
+    "string2", [] (const IValue& v) {EXPECT_EQ("string2", v.toString()->string());});
+  ArgTypeTestKernel<std::string>::test(
+    "string1", [] (const std::string& v) {EXPECT_EQ("string1", v);},
+    "string2", [] (const IValue& v) {EXPECT_EQ("string2", v.toString()->string());},
+    "(str a) -> str");
+  ArgTypeTestKernel<Tensor>::test(
+    dummyTensor(TensorType1()), [] (const Tensor& v) {EXPECT_EQ(TensorType1(), v.type_id());},
+    dummyTensor(TensorType2()), [] (const IValue& v) {EXPECT_EQ(TensorType2(), v.toTensor().type_id());});
+  ArgTypeTestKernel<Tensor>::test(
+    dummyTensor(TensorType1()), [] (const Tensor& v) {EXPECT_EQ(TensorType1(), v.type_id());},
+    dummyTensor(TensorType2()), [] (const IValue& v) {EXPECT_EQ(TensorType2(), v.toTensor().type_id());},
+    "(Tensor a) -> Tensor");
+
+
+  // optional types (with has_value() == true)
```

Comment:
Is it really necessary to test every optional permutation this way?

Reply:
better be safe than sorry. These are different argument types and I want to make sure they work. As an example: Only by doing all permutations for lists (i.e. `int[]`, `float[]`, ...) did I notice that `str[]` actually does not work and isn't implemented.